## ※ 공분산?  

확률론과 통계학에서 2개의 확률변수의 상관정도를 나타내는 값이다.


### ■ 키와 체중간의 상관관계가 있는지 확인하고 싶다면?


### ■ 공분산

In [8]:
tall = [147.9, 163.5, 159.8, 155.1, 163.3, 158.7, 172.0, 161.2, 153.9, 161.6]
weight = [41.7, 60.2, 47.0, 53.2, 48.3, 55.2, 58.5, 49.0, 46.7, 52.5]


sum1 = 0
for i in tall:
    sum1 += i

x_bar = sum1/len(tall)
print(x_bar)

sum2 = 0
for i in weight:
    sum2 += i

y_bar = sum2/len(weight)
print(y_bar)    

total = 0
for i in range(0,len(tall)):
   total += (tall[i]-x_bar)*(weight[i]-y_bar)

print(total)
var = total/len(tall)
print(var)

#결과 : 공분산 값은 23.7입니다.
#키가 크면 체중도 늘어난다고 하는 양의 상관관계를 증명!

159.7
51.23
237.29999999999976
23.729999999999976


## 예제72. 최대값, 최소값 구하기(max,min)

max()와 min()은 인자로 입력된 자료에서 최대, 최소값을 구해주는함수입니다.  
max()와 min()의 인자로 리스트와 같은 시퀀스 자료가 입력되면   
시퀀스 자료의 요소 가운대 최대값 또는 최소값을 구하여 리턴합니다.  

| 오라클 | 파이썬 |
|:------:|:------:|
|   max  |   max  |
|   min  |   min  |
|  count |   len  |
|   sum  |   sum  |
|   avg  |  mean  |

In [9]:
listdata = [8,7,12,55,21,34,15,9,22]
result = max(listdata)
print(result)

55


## 문제143. 사원 테이블에서 최대 월급을 출력하시오!

### 1. pandas를 이용한 방법

### 2. pandas를 이용하지 않는 방법  

In [10]:
import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")
print (emp['sal'].max())

5000


In [15]:
#pandas 이용하지 않는 방법 (컬럼명이 없는 csv를 가져와야됨.)

import csv
file = open("d:\\data\emp2.csv")
emp = csv.reader(file)
a=[]
for i in emp:
    a.append(int(i[5])) #문자형으로 나타나지기 때문에 int로 숫자형으로 바꾸어 주어야됨!

print(a)
result = max(a)
print(result)

[5000, 2850, 2450, 2975, 1250, 1600, 1500, 950, 1250, 3000, 800, 3000, 1100, 1300]
5000


## 문제144. 아래의 SQL을 판다스로 구현하시오

SQL>

select deptno, max(sal)  
    from emp  
    group by deptno;  

In [17]:
import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")
result = emp.groupby('deptno')['sal'].max().reset_index()
print(result)

   deptno   sal
0      10  5000
1      20  3000
2      30  2850


## 문제145. 부서번호, 직업, 부서번호별 직업별 최대 월급을 출력하시오!


select deptno, job, max(sal)  
    from emp  
    group by deptno, job  
    order by deptno, job  

In [19]:
import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")
result = emp.groupby(['deptno','job'])['sal'].max().reset_index()
print(result)

   deptno        job   sal
0      10      CLERK  1300
1      10    MANAGER  2450
2      10  PRESIDENT  5000
3      20    ANALYST  3000
4      20      CLERK  1100
5      20    MANAGER  2975
6      30      CLERK   950
7      30    MANAGER  2850
8      30   SALESMAN  1600


## 문제146. 통신사, 통신사별 최대나이를 출력하시오!(pandas로)


In [3]:
import pandas as pd
emp = pd.read_csv("d:\\data\emp10.csv")
result = emp.groupby('telecom')['age'].max().reset_index()
print(result)

  telecom  age
0      kt   34
1      lg   40
2     skt   34


## Tip.

### Character set(문자를 담는 방)

   1. __English__ : ANSI 문자방 

   2. __한글__ : UTF-8(유니코드 문자셋) 

## 예제73. 복잡한 쿼리문을 판다스로 구현하기!

아래의 SQL을 판다스로 구현하시오~

In [16]:
"""
SQL>
select job, sum(sal)
    from emp
    group by job
    having sum(sal)>= 6000;
"""

#Pandas

import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")

result = emp.groupby('job')['sal'].sum().reset_index()
print(type(result))

result2 = result[['job','sal']] [result['sal'] >= 6000] 
# reset_index()을 해줘야 type이 DataFrame으로 나타나져서 result2의 결과를 나타 낼 수있다.

print(type(result2))

print(result2)

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
       job   sal
0  ANALYST  6000
2  MANAGER  8275


| 오라클 |   판다스  |
|:------:|:---------:|
| 테이블 | DataFrame |
|  컬럼  |   Series  |

## 문제147. 아래의 SQL을 판다스로 구현하시오!

In [22]:
"""
SQL>
select job,sum(sal)
    from emp
    group by job
    having sum(sal) >= 6000
    order by sum(sal) desc;
    
"""

#Pandas

import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")

result = emp.groupby('job')['sal'].sum().reset_index()

print(result)

result2 = result[['job','sal']] [result['sal'] >= 6000] 

result3 = result2.sort_values('sal', ascending = False )

print(result3)

         job   sal
0    ANALYST  6000
1      CLERK  4150
2    MANAGER  8275
3  PRESIDENT  5000
4   SALESMAN  5600
       job   sal
2  MANAGER  8275
0  ANALYST  6000


## 문제148. 아래의 SQL을 판다스로 구현하시오!

In [23]:
"""
SQL>
select job, sum(sal)
    from emp
    where job != 'SALESMAN'
    group by job
    having sum(sal) >= 6000
    order by sum(sal) desc;
"""


# Pandas

import pandas as pd
emp = pd.read_csv("d:\\data\emp3.csv")

result = emp.groupby('job')['sal'].sum().reset_index()

result1 = result[['job','sal']][result['job'] != 'SALESMAN']
print(result1)

result2 = result1[['job','sal']] [result1['sal'] >= 6000] 

result3 = result2.sort_values('sal', ascending = False )

print(result3)


# from -> where -> group by - > having -> selct -> order by

         job   sal
0    ANALYST  6000
1      CLERK  4150
2    MANAGER  8275
3  PRESIDENT  5000
       job   sal
2  MANAGER  8275
0  ANALYST  6000


## 예제74. 판다스와 오라클 분석함수 비교

### 순위 출력하는 분석함수 : rank

| 오라클 | 판다스 |
|:------:|:------:|
| rank() | rank() |

In [ ]:
"""
SQL>
select ename, sal, rank() over (order by sal desc) 순위
    from emp;
"""

# Pandas

import pandas as pd

emp = pd.read_csv("d:\\data\emp3.csv")

emp['rnk'] = emp['sal'].rank(ascending = False).apply(int)

emp = emp.sort_values('rnk', ascending = True)

print(emp[['ename','sal','rnk']])

## 문제149. 아래의 SQL을 판다스로 구현하시오!

In [ ]:
"""
SQL>
select job, sum(sal), rank() over (order by sumsal desc) 순위
    from (select job, sum(sal) as sumsal
            from emp
            group by job);
"""

# Pandas

import pandas as pd

emp = pd.read_csv("d:\\data\emp3.csv")

empsub = emp.groupby('job')['sal'].sum().reset_index()

print(empsub)

empsub['rnk'] = empsub['sal'].rank(ascending = False).apply(int)
empsub = empsub.sort_values('rnk', ascending = True)

print(empsub[['job','sal','rnk']])

In [36]:
import pandas as pd

emp = pd.read_csv("d:\\data\emp3.csv")

empsub = emp.groupby('job')['sal'].sum().reset_index()

print(empsub)

empsub['rnk'] = empsub['sal'].rank(ascending = False).apply(int)


print(empsub[['job','sal','rnk']].sort_values('rnk', ascending = True))

         job   sal
0    ANALYST  6000
1      CLERK  4150
2    MANAGER  8275
3  PRESIDENT  5000
4   SALESMAN  5600
         job   sal  rnk
2    MANAGER  8275    1
0    ANALYST  6000    2
4   SALESMAN  5600    3
3  PRESIDENT  5000    4
1      CLERK  4150    5


## 문제150. 아래 SQL을 Pandas로 구현하시오~

In [41]:
"""
SQL>
select telecom, maxage, rank() over (order by maxage desc) 순위
    from
    (select telecom, max(age) as maxage
        from emp10
        group by telecom
        );
"""
# Pandas

import pandas as pd
emp = pd.read_csv("d:\\data\emp10.csv")

empsub = emp.groupby('telecom')['age'].max().reset_index()

empsub['rnk'] = empsub['age'].rank(ascending = False).apply(int)

print(empsub[['telecom','age','rnk']].sort_values('rnk', ascending = True))

  telecom  age  rnk
1      lg   40    1
0      kt   34    2
2     skt   34    2


## 예제75. 문자열에서 특정위치의 문자 얻기

문자열에서 특정 위치의 문자를 얻는 방법은 인덱싱을 이용하는 것입니다.  
인덱스는 0부터 시작합니다. 파이썬 인덱스는 음수도 가능합니다.

In [42]:
txt1 = 'A tale that was not right'
print(txt1[5])

e


## 문제151. 위의 txt1 변수에서 끝의 철자 g를 출력하시오!

In [43]:
txt1 = 'A tale that was not right'
print(txt1[-3])

g


## 예제75. 문자열에서 지정한 구간의 문자열 얻기

문자열에서 특정 구간에 있는 문자열을 얻으려면 슬라이싱을 이용하면 된다.

In [45]:
txt1 = 'A tale that was not right'
print(txt1[2:8]) #2부터 8미만까지의 문자열을 출력
print(txt1[2:]) #2부터 문자 끝까지 출력
print(txt1[:8]) #문자열 처음부터 8미만까지 출력


tale t


## 문자151. 위의 슬라이싱 문법을 이용해서 아래의 txt변수의 문자열을 아래와 같이 출력하시오!

txt = 'python'

결과:  
p  
py  
pyt  
pytho  
python  


In [53]:
txt = 'python'

for i in range(1,7):
    print(txt[0:i])

p
py
pyt
pyth
pytho
python


In [54]:
txt = 'python'

for i in range(len(txt)):
    print(txt[:i+1])

p
py
pyt
pyth
pytho
python


## 예제77. 문자열에서 홀수 번째 문자만 출력하기

주어진 문자열에서 홀수 번째 문자만 추출하는 방법은 슬라이싱의 스탭을 이용하면 간단해진다.

In [56]:
txt = 'aAbBcCdDeEfFgGhHiIjJkK'
result = txt[::2] # 마지막에 있는 2가 step!!
print(result) 

abcdefghijk


## 문제152. 문자열 txt에서 짝수번째 철자만 추출하시오!

In [57]:
txt = 'aAbBcCdDeEfFgGhHiIjJkK'
result = txt[1::2] # 마지막에 있는 2가 step!!
print(result) 

ABCDEFGHIJK


## 예제78. 문자열 거꾸로 만들기

슬라이싱을 이용하면 매우 간단하게 거꾸로 된 문자열을 얻을 수 있다.  
문자열 txt의 처음부터 끝까지 스텝 -1로 슬라이싱을 하면 된다.

In [58]:
txt = 'aAbBcCdDeEfFgGhHiIjJkK'
result = txt[::-1] 
print(result) 

KkJjIiHhGgFfEeDdCcBbAa


## 문제153. txt 문자열을 거꾸로 출력하는데 홀수 번째 문자만 출력해보시오!

In [63]:
txt = 'aAbBcCdDeEfFgGhHiIjJkK'
result = txt[::-2] 
print(result) 

KJIHGFEDCBA


## 문제154. txt 문자열을 거꾸로 출력하는데 짝수 번째 문자만 출력하시오!

In [1]:
txt = 'aAbBcCdDeEfFgGhHiIjJkK'
result = txt[-2::-2] 
print(result) 

kjihgfedcba


## 예제79. 두개의 문자열 합치기(+)

두개의 문자열을 합치는 방법은 매우 간단하다.  
문자열1 + 문자열2 + 문자열3으로 + 연산자를 이용하면 된다.  

In [67]:
print('재훈형' + '베리' + '핸섬')

재훈형베리핸섬


## 예제80. 문자열을 반복해서 새로운 문자열로 만들기(*)

주어진 문자열을 반복하고자 할 때 * 연산자를 이용한다.


In [66]:
print('여러분~ '*3)

여러분~ 여러분~ 여러분~ 


## 예제81. 문자열에서 특정 문자가 있는지 확인하기!(in)

문자열에서 특정 문자가 있는지 없는지 확인하려면 in 키워드를 이용하면 된다.

In [ ]:
msg = 'abcdefghijklmn'

if 'b' in msg:
    print('존재합니다.')
else:
    print('존재하지 않습니다.')

In [68]:
msg = 'I am a boy'
if 'boy' in msg:
    print('존재합니다.')
else:
    print('존재하지 않습니다.')

존재합니다.


## 예제83. 파이썬 comprehension

파이썬 Comprehension을 이용해서 파이썬 코드를 더 간결하고 심플하게 작성 할 수 있습니다.

＊ Python comprehension 코드의 종류 3가지

1. list comprehension
2. set comprehension
3. dictionary comprehension


## 예제 : 아래 a리스트에서 숫자만 추출해서 리스트에 담아 출력하시오!

In [73]:
# comprehension 안한 코드

a = [1, 2, 'A', False, 3]
b = []
for i in a:
    if type(i) == int:
        b.append(i)
        
print(b)

[1, 2, 3]


In [75]:
# comprehension 한 코드

a = [1, 2, 'A', False, 3]
b = [i for i in a  if type(i) == int]

print(b)

[1, 2, 3]


### comprehension 문법표현 

___출력표현식 for 요소 in 입력시퀀스 if 조건식 (그냥 b보고 습득하는것을 추천)___

## 문제155. list comprehension을 이용해서 아래의 결과를 출력하시오!

[1,3,5,7,9,11,13,15,17,19]

In [ ]:
a = []
for i in range(0,21):
    a.append(i)
    
print(a)

b = [i for i in a if i%2 == 1]
print(b)

In [82]:
print([2*i + 1 for i in range(10)])

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]


## 문제156. emp2.csv에서 최대월급을 출력하는 아래의 코드를 comprehension 코드로 변경하시오!

In [86]:
import csv
file = open("d:\\data\emp2.csv", "r")
emp_csv = csv.reader(file)
emp_sal = []

for i in emp_csv:
    emp_sal.append(int(i[5]))

print(max(emp_sal))



5000


In [ ]:
"""
emp_sal = []

for i in emp_csv:
    emp_sal.append(int(i[5]))
    
## 위의 3줄을 list comprehension을 이용하여 한줄로 만들어보기!##
""""""

In [89]:
import csv
file = open("d:\\data\emp2.csv", "r")

emp_csv = csv.reader(file)

emp_sal = [int(i[5]) for i in emp_csv]

print(max(emp_sal))

5000


## ＊ set comprehension

문법 : { 출력표현식 for 요소 in 입력시퀀스 if 조건식 }

## 예제 : 아래의 리스트의 데이트를 중복 제거해서 출력하시오!

a = [1,1,2,2,3,3,3,4]    ---->   {1,2,3,4}

In [91]:
# Comprehension을 안한 코드
a = [1,1,2,2,3,3,3,4]
b = []

for i in a:
    b.append(i)    
b2 = set(b) # set을 이용하여 중복을 제거 할 수 있다.
print(b2)

{1, 2, 3, 4}


In [1]:
"""
b = []

for i in a:
    b.append(i)    
b2 = set(b)
"""
# 이 4줄을 comprehension을 이용하여 1줄로 만드는 방법.
# comprehension 을 사용한 코드

a = [1,1,2,2,3,3,3,4]
b = {i for i in a}

print(b)
print(type(b))
c = list(b)
print(c)

{1, 2, 3, 4}
<class 'set'>
[1, 2, 3, 4]


## ＊ 딕셔너리 comprehension

{ key:Value for 요소 in 입력시퀀스 if 조건식 }

In [94]:
id_name = {1:'김인엽', 2:'최인선', 3:'김재훈', 4:'노민준'}
print(type(id_name))
print(id_name)
print(id_name.keys())
print(id_name.values())
print(id_name.items()) # dict_items 는 key 와 value가 같이 나온다.

<class 'dict'>
{1: '김인엽', 2: '최인선', 3: '김재훈', 4: '노민준'}
dict_keys([1, 2, 3, 4])
dict_values(['김인엽', '최인선', '김재훈', '노민준'])
dict_items([(1, '김인엽'), (2, '최인선'), (3, '김재훈'), (4, '노민준')])


## 예제 : id_name에 있는 데이터를 a라는 딕셔너리로 구성하시오!

In [100]:
# comprehension을 안 쓴 코드
# id_name 딕셔너리를 a 딕셔너리에 옮기는 과정

id_name = {1:'김인엽', 2:'최인선', 3:'김재훈', 4:'노민준'}
a = {}
for key,val in id_name.items():
    a[key] = val
print(a)

{1: '김인엽', 2: '최인선', 3: '김재훈', 4: '노민준'}


In [99]:
# comprehension 사용한 코드

"""
a = {}
for key,val in id_name.items():
    a[val] = key
print(a)

이부분을 간단하게 표현할 수 있다.
"""

id_name = {1:'김인엽', 2:'최인선', 3:'김재훈', 4:'노민준'}
a = { key : val for key, val in id_name.items()}
print(a)

{1: '김인엽', 2: '최인선', 3: '김재훈', 4: '노민준'}


## 예제84. 문자열이 알파벳인지 검사하기(isalpha)

문자열은 문자나 숫자, 기호들로 구성됩니다.  
코드를 작성하다 보면 특정 문자열이 한글이나 알파벳과 같이  
사람의 언어를 표현하기 위해 사용되는 문자로만 구성되어 있는지  
확인해야할 경우가 있다.  

파이썬 문자열 객체가 제공하는 메소드인 isalpha()는 주어진 문자열이 사람의 언어 문자로만 구성되어 있는지 확인해준다.

In [ ]:
tx1 = 'Warcraft three'
txt2 = '안녕'
txt3 = '3PO'

print (txt1.isalpha()) # 공백때문에 False
print (txt2.isalpha()) # True
print (txt3.isalpha()) # 숫자때문에 False

## 문제157. 겨울왕국 대본에서 알파벳 문자의 갯수가 어떻게 되는가?

In [ ]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines() # 텍스트 파일을 한줄 씩 읽기
print(lines) # 하나의 리스트에 겨울왕국 대본 전체가 들어가 있는 상황

In [ ]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines()


for i in lines:
    print(i) # 한줄씩 나누어서 print 하기

In [ ]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines()


for i in lines:
    for k in i:
        print(k) # 한줄씩 읽은 뒤 철자 하나씩 표현

In [107]:
# 최종답

text_file = open("d:\\data\winter.txt")
lines = text_file.readlines() # 텍스트 파일을 한줄 씩 읽기
cnt = 0

for i in lines: 
    for k in i:
        if k.isalpha() == True:
            cnt += 1
            
print(cnt)

82180


In [ ]:
"""
for k in i:
    if k.isalpha() == True:
        cnt += 1
        
을 comprehension으로 만들어보기
"""
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines() # 텍스트 파일을 한줄 씩 읽기
total = 0

for s in lines: 
    cnt = sum(k.isalpha() for k in s) # True = 1, False = 0
    print(s,cnt)
    total = total + cnt
    
print(total)

## 예제85. 문자열이 숫자인지 검사하기(isdigit)

문자열 객체의 isdigit() 메소드는 문자열을 구성하는 요소가  
모두 숫자인지 체크하고 True 또는 False를 리턴합니다.  

| 번호 |  내장함수 |            설명            |
|:----:|:---------:|:--------------------------:|
|   1  | isalpha() | 알파벳(한글)이 맞는지 확인 |
|   2  | isdigit() |     숫자가 맞는지 확인     |
|   3  | isspace() |     공백이 맞는지 확인     |

## 문제159. 겨울왕국에 숫자가 몇개가 있는지 출력하시오!

In [ ]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines() # 텍스트 파일을 한줄 씩 읽기
total = 0

for s in lines: 
    cnt = sum(k.isdigit() for k in s)
    print(s,cnt)
    total = total + cnt
    
print(total)

## ■ 문자열이 특수문자인지 검사하기

#### ★ 특수문자 = 전체문자 - (문자 + 숫자 + 공백)


__전체문자__ = 문자, 숫자, 공백, 특수문자 전부 다 (all)

## 문제160. 겨울왕국 대본의 전체 문자의 갯수가 어떻게 되는지?

In [126]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines()
total1 = 0

for s in lines:
    cnt = len(s)
    total1 = total1 + cnt
    
print(total1)


109438


## 문제161. 겨울왕국 대본에서 특수문자가 몇개가 있는지 확인하시오!

특수문자 = 전체문자 -(문자 + 숫자 + 공백)

In [9]:
text_file = open("d:\\data\winter.txt")
lines = text_file.readlines() # 텍스트 파일을 한줄 씩 읽기
total1 = 0
total2 = 0
total3 = 0
total4 = 0

for s in lines: 
    cnt1 = sum(k.isalpha() for k in s)
    cnt2 = sum(k.isdigit() for k in s)
    cnt3 = sum(k.isspace() for k in s)
    cnt4 = len(s)
    
    
    total1 = total1 + cnt1
    total2 = total2 + cnt2
    total3 = total3 + cnt3
    total4 = total4 + cnt4
    
print(total1)
print(total2)
print(total3)
print(total4)

print(total4 - total1 - total2 -total3)

82180
264
19911
109438
7083


In [10]:
text_file = open("d:\\datadata\winter.txt")
lines = text_file.readlines()   # 텍스트 파일을 한줄 씩 읽기

cnt2 = 0
for i in lines:
    cnt2 += sum(1 for k in i if not ( k.isalpha() or k.isdigit() or k.isspace() ) )    
        
print(cnt2)

FileNotFoundError: [Errno 2] No such file or directory: 'd:\\datadata\\winter.txt'

## 오늘의 마지막 문제 :   

### 이진탐색 스터디를 다 한뒤에 이진 탐색을 이해하고 아래의 리스트에서 문자 h를 이진 탐색으로 검색하시오.

아래의 리스트 a에서 67을 검색하는 방법을 이진탐색으로 알려줄 예정.  

a = [1,7,11,12,14,23,33,47,51,64,67,77,139,672,871]  

이진탐색(index scan)이 순차탐색(full table scan)보다 훨씬 검색속도가 빠르다.

In [135]:
"""
SQL>
select ename, sal
    from emp
    where ename = 'SCOTT';
"""
b = ['a','b','c','d','e','f','g','h','i','j','k','i','m','n','o','p','q','r','s']

print(ord('a')) # 알파벳 a에 대한 아스키 코드
print(ord('b'))
print(ord('s'))

#구글에서 아스키 코드표를 입력해볼것

97
98
115


In [136]:
a = []
b = ['a','b','c','d','e','f','g','h','i','j','k','i','m','n','o','p','q','r','s']

for i in range(0,len(b)):
        a.append(ord(b[i]))

print(a)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 105, 109, 110, 111, 112, 113, 114, 115]


In [7]:
#%% 위의 답을 def로 정의해보기 
# Tip. 컴퓨터는 문자끼리 부등호로 비교할 때 자동으로 아스키코드로 비교한다.
# 따라서 굳이 ord를 쓸 필요까지는 없다.
# 문자도 sort하면 아스키 코드값대로 순서대로 정렬된다.
def b_search(b):
    try:
        d = input('검색할 문자를 입력하세요~')
        for i in range(1,len(b)): 
        # range범위는 큰 의미없음 그냥넉넉하게 반복문 돌리려고하는거임.
        # 따라서 while true: 로 무한루프 돌려도 됨.
            
            half = int(len(b)/2) #리스트의 중앙값의 위치 len(a)/2에다가 int로 정수형만들기
            
            if d > b[half]: # 찾고자 하는 값이 중앙값보다 클 때
                del b[0:half+1] #처음부터 중간까지 cut
            
            elif d < b[half]: # 찾고자 하는 값이 중앙값보다 클 때
                del b[half:] # 중간부터 끝까지 cut
            
            else: # d = b[half] 찾고자 하는 값이 중앙값이랑 같을 떄
                del b[0:half] #왼쪽 사이드 다 cut
                del b[1:] #오른쪽 사이드 다 cut
                break
        return b
    except:
        return print('해당 문자가 없습니다.')

b = ['a','b','c','d','e','f','g','h','i','j','k','i','m','n','o','p','q','r','s']
print(b_search(b))


검색할 문자를 입력하세요~h
['h']
